test/hotspot/jtreg/compiler/c2/TestAbs.java

```diff
@@ -21,17 +21,151 @@
+        Float.NEGATIVE_INFINITY,
+        Integer.MAX_VALUE,
+        Integer.MIN_VALUE,
+        Long.MAX_VALUE,
+        Long.MIN_VALUE,
+        -1.0f,
+        1.0f,
+        -42.0f,
+        42.0f
+    };
+
+    public static double [] dspecial = {
+        0.0,
+        -0.0,
+        Double.MAX_VALUE,
+        Double.MIN_VALUE,
+        -Double.MAX_VALUE,
+        -Double.MIN_VALUE,
+        Double.NaN,
+        Double.POSITIVE_INFINITY,
+        Double.NEGATIVE_INFINITY,
+        Integer.MAX_VALUE,
+        Integer.MIN_VALUE,
+        Long.MIN_VALUE,
+        Long.MAX_VALUE,
+        -1,
+        1,
+        42,
+        -42,
+        Math.PI,
+        Math.E,
+        Float.MAX_VALUE,
+        Float.MIN_VALUE,
+        -Float.MAX_VALUE,
+        -Float.MIN_VALUE,
+        Float.NaN,
+        Float.POSITIVE_INFINITY,
+        Float.NEGATIVE_INFINITY
+    };
+
+    public static void testAbsConstant() {
+        // Test abs(constant) optimization for int
+        Asserts.assertEquals(Integer.MAX_VALUE, Math.abs(Integer.MAX_VALUE));
+        Asserts.assertEquals(Integer.MIN_VALUE, Math.abs(Integer.MIN_VALUE));
+        Asserts.assertEquals(Integer.MAX_VALUE, Math.abs(-Integer.MAX_VALUE));
+
+        // Test abs(constant) optimization for long
+        Asserts.assertEquals(Long.MAX_VALUE, Math.abs(Long.MAX_VALUE));
+        Asserts.assertEquals(Long.MIN_VALUE, Math.abs(Long.MIN_VALUE));
+        Asserts.assertEquals(Long.MAX_VALUE, Math.abs(-Long.MAX_VALUE));
+
+        // Test abs(constant) optimization for float
+        Asserts.assertEquals(Float.NaN, Math.abs(Float.NaN));
+        Asserts.assertEquals(Float.POSITIVE_INFINITY, Math.abs(Float.NEGATIVE_INFINITY));
+        Asserts.assertEquals(Float.POSITIVE_INFINITY, Math.abs(Float.POSITIVE_INFINITY));
+        Asserts.assertEquals(0.0f, Math.abs(0.0f));
+        Asserts.assertEquals(0.0f, Math.abs(-0.0f));
+        Asserts.assertEquals(Float.MAX_VALUE, Math.abs(Float.MAX_VALUE));
+        Asserts.assertEquals(Float.MIN_VALUE, Math.abs(Float.MIN_VALUE));
+        Asserts.assertEquals(Float.MAX_VALUE, Math.abs(-Float.MAX_VALUE));
+        Asserts.assertEquals(Float.MIN_VALUE, Math.abs(-Float.MIN_VALUE));
+
+        // Test abs(constant) optimization for double
+        Asserts.assertEquals(Double.NaN, Math.abs(Double.NaN));
+        Asserts.assertEquals(Double.POSITIVE_INFINITY, Math.abs(Double.NEGATIVE_INFINITY));
+        Asserts.assertEquals(Double.POSITIVE_INFINITY, Math.abs(Double.POSITIVE_INFINITY));
+        Asserts.assertEquals(0.0, Math.abs(0.0));
+        Asserts.assertEquals(0.0, Math.abs(-0.0));
+        Asserts.assertEquals(Double.MAX_VALUE, Math.abs(Double.MAX_VALUE));
+        Asserts.assertEquals(Double.MIN_VALUE, Math.abs(Double.MIN_VALUE));
+        Asserts.assertEquals(Double.MAX_VALUE, Math.abs(-Double.MAX_VALUE));
+        Asserts.assertEquals(Double.MIN_VALUE, Math.abs(-Double.MIN_VALUE));
+    }
+
+    private static void testAbsTransformInt(int[] a) {
```

Comment:
If you want to verify C2's transformation, probably we should use C2's IR test framework.

Reply:
Done. Thanks.